openqa-bats-review

```diff
@@ -0,0 +1,185 @@
+
+
+# We want to extract the test name stripping the numbers and the optional timing information
+# not ok 166 bud-git-context in 118ms
+# not ok 655 [520] podman checkpoint --export, with volumes in 1558ms
+NOT_OK = re.compile(r"^not ok \d+ (?:\[\d+\] )?(.*)(?: in \d+ms)?$")
+
+
+def get_file(url: str) -> str:
+    """
+    Get a text file
+    """
+    headers = {
+        "User-Agent": USER_AGENT,
+    }
+    try:
+        got = session.get(url, headers=headers, timeout=TIMEOUT)
+        got.raise_for_status()
+    except RequestException as error:
+        log.error("%s: %s", url, error)
+        sys.exit(1)
+    return got.text
+
+
+@cache
+def get_job(url: str) -> dict:
+    """
+    Get a job
+    """
+    headers = {
+        "User-Agent": USER_AGENT,
+    }
+    try:
+        got = session.get(url, headers=headers, timeout=TIMEOUT)
+        got.raise_for_status()
+        data = got.json()
+    except RequestException as error:
+        log.error("%s: %s", url, error)
+        sys.exit(1)
+    return data["job"]
+
+
+def grep_notok(url: str) -> set[str]:
+    """
+    Grep for "not ok" lines and return a set with the failing tests
+    prefixed by the filename
+    """
+    notok = set()
+    prefix = os.path.basename(url)
+    data = get_file(url)
+    lines = data.splitlines()
+    last = 0
+    for line in lines:
+        if line.startswith("1.."):
+            last = int(line.split("..", 1)[1])
+            break
+    if not last:
+        log.error("Malformed TAP file: %s", url)
+        sys.exit(1)
+    tests = 0
+    for line in lines:
+        if not line.startswith(("ok", "not ok", "#not ok")):
+            continue
+        tests += 1
+        try:
+            test = NOT_OK.findall(line)[0]
+            notok.add(f"{prefix}:{test}")
+        except IndexError:
+            continue
+    if tests != last:
+        log.error("Truncated TAP file: %s", url)
+        sys.exit(1)
+    return notok
+
+
+def process_files(files: list[str]) -> set[str]:
+    """
+    Process TAP files
+    """
+    with ThreadPoolExecutor(max_workers=len(files)) as executor:
+        return set(itertools.chain.from_iterable(executor.map(grep_notok, files)))
+
+
+def get_clone_chain(openqa_host: str, job_id: int) -> list[int]:
+    """
+    Follow clones recursively and return the full chain:
+    [job_id, origin_id, origin_id_of_origin, ...]
+    """
+    chain = []
+    current = job_id
+    while current:
+        # We use "/details" because we'll need this information again and get_job() is cached
+        job = get_job(f"{openqa_host}/api/v1/jobs/{current}/details")
+        if "BATS_PACKAGE" not in job["settings"]:
+            log.error("Not a BATS test: %d", job_id)
+            sys.exit(1)
+        chain.append(current)
+        current = job.get("origin_id")
+    return chain
+
+
+def main(url: str) -> None:
+    """
+    Main function
+    """
+    if not url.startswith(("http://", "https://")):
+        url = f"https://{url}"
```

Comment:
you already use urlparse. so it would make sense to move `urlx = urlparse(url)` above and use `urlx.scheme`. wdyt?

Reply:
Maybe
```
In [1]: from urllib.parse import urlparse

In [2]: urlparse("openqa.suse.de/tests/123")
Out[2]: ParseResult(scheme='', netloc='', path='openqa.suse.de/tests/123', params='', query='', fragment='')

In [3]: u=urlparse("openqa.suse.de/tests/123")

In [4]: all([u.scheme,u.netloc])
Out[4]: False

In [5]: u=urlparse("http://openqa.suse.de/tests/123")

In [6]: all([u.scheme,u.netloc])
```

Reply:
I think the current check is fine